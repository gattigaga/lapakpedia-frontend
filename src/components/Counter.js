import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  width: 96px;
  height: 32px;
  border: 1px solid black;
  display: flex;
`;

export const Button = styled.button`
  background: black;
  color: white;
  width: 32px;
  height: 100%;
  border: 0px;
  font-family: Roboto;
  font-size: 18px;
  cursor: pointer;
  outline: none;
  user-select: none;
`;

export const Input = styled.input`
  width: 32px;
  height: 100%;
  box-sizing: border-box;
  font-family: Roboto;
  font-size: 16px;
  border: 0px;
  text-align: center;
  outline: none;
`;

const Counter = ({ value, onClickPlus, onClickMinus, onChange }) => (
  <Container>
    <Button onClick={onClickMinus}>-</Button>
    <Input type="text" value={value} maxLength={2} onChange={onChange} />
    <Button onClick={onClickPlus}>+</Button>
  </Container>
);

Counter.propTypes = {
  onClickPlus: PropTypes.func,
  onClickMinus: PropTypes.func,
  onChange: PropTypes.func,
  value: PropTypes.number
};

Counter.defaultProps = {
  value: 0
};

export default Counter;
