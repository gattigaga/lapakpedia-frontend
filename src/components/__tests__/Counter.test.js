import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Counter, { Input, Button } from "components/Counter";

describe("Counter", () => {
  const setup = propOverrides => {
    const props = {
      ...propOverrides
    };

    const wrapper = shallow(<Counter {...props} />);

    return {
      props,
      wrapper
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders with custom value", () => {
    const { wrapper } = setup({ value: 10 });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders with max value", () => {
    const { wrapper } = setup({ value: 25, maximum: 25 });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should calls 'onClickPlus' callback while plus button clicked", () => {
    const onClickPlus = jest.fn();
    const { wrapper } = setup({ onClickPlus });

    wrapper
      .find(Button)
      .at(1)
      .simulate("click");

    expect(onClickPlus).toBeCalled();
  });

  it("should calls 'onClickMinus' callback while minus button clicked", () => {
    const onClickMinus = jest.fn();
    const { wrapper } = setup({ onClickMinus });

    wrapper
      .find(Button)
      .at(0)
      .simulate("click");

    expect(onClickMinus).toBeCalled();
  });

  it("should calls 'onChange' callback while input box value was changed", () => {
    const onChange = jest.fn();
    const { wrapper } = setup({ onChange });
    const event = {
      target: {
        value: 99
      }
    };

    wrapper.find(Input).simulate("change", event);

    expect(onChange).toBeCalledWith(event);
  });
});
