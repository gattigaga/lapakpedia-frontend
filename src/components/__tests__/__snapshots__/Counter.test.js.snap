// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should renders default 1`] = `
<styled.div>
  <styled.button
    disabled={true}
  >
    -
  </styled.button>
  <styled.input
    maxLength={2}
    type="text"
    value={1}
  />
  <styled.button>
    +
  </styled.button>
</styled.div>
`;

exports[`Counter should renders with custom value 1`] = `
<styled.div>
  <styled.button
    disabled={false}
  >
    -
  </styled.button>
  <styled.input
    maxLength={2}
    type="text"
    value={99}
  />
  <styled.button>
    +
  </styled.button>
</styled.div>
`;
